### PR TITLE
Handle new envoy-resources and reassignment before expiration

### PR DESCRIPTION
**Pull request by itzg**
# Resolves

Part of
- https://jira.rax.io/browse/SALUS-263
- https://jira.rax.io/browse/SALUS-311

# What

This introduces just enough of the zone management application to
- watch when a new envoy-resource registers into the expected zone key range
- watch when a pre-existing envoy-resource re-registers with the same resource ID
- produces a corresponding zone event for each scenario

Detailed sequence diagrams were created to assist me in developing the code for this and corresponding monitor management changes (soon to be PR'ed):

https://one.rackspace.com/pages/viewpage.action?pageId=577472836#ExplicitPollerPickingDesign-DetailedSequenceDiagrams

# Depends on

- https://github.com/racker/salus-telemetry-etcd-adapter/pull/29

## How to test

New unit tests were created.

### CODE REVIEW

File: src/main/java/com/rackspace/salus/zone_mgmt/services/ZoneWatchingService.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    this.zoneStorage = zoneStorage;
+    this.kafkaTemplate = kafkaTemplate;
+    this.topicProperties = topicProperties;
+  }
+
+  @PostConstruct
+  public void start() {
+
+    zoneStorage.watchExpectedZones(this)
+        .thenAccept(watcher -> {
```
Comment (harageth) [inline src/main/java/com/rackspace/salus/zone_mgmt/services/ZoneWatchingService.java]: What is the watcher in this instance?

Comment (itzg) [reply src/main/java/com/rackspace/salus/zone_mgmt/services/ZoneWatchingService.java]: The jetcd watcher object completed via this returned future:

https://github.com/racker/salus-telemetry-etcd-adapter/pull/29/files#diff-84ecb523c8a049bf1f41970434b9df74R216

I retain it there so I can gracefully close it when Spring stops this outer service.

Comment (harageth) [reply src/main/java/com/rackspace/salus/zone_mgmt/services/ZoneWatchingService.java]: Ahhh right ok that makes sense

File: src/main/java/com/rackspace/salus/zone_mgmt/ZoneManagementApplication.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+package com.rackspace.salus.zone_mgmt;
+
+import com.rackspace.salus.common.messaging.EnableSalusKafkaMessaging;
+import com.rackspace.salus.telemetry.etcd.EnableEtcd;
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+
+@SpringBootApplication
+@EnableSalusKafkaMessaging
+@EnableEtcd
```
Comment (jjbuchan) [inline src/main/java/com/rackspace/salus/zone_mgmt/ZoneManagementApplication.java]: Hmm, we don't seem to use this in various repos.  Seems like it should be added to almost all of them?

Comment (itzg) [reply src/main/java/com/rackspace/salus/zone_mgmt/ZoneManagementApplication.java]: It's used on all the ones that need it...except presence monitor. That must have been implemented concurrently with me adding this helper bootstrap config.

**Pull request merged**